### ISSUE TITLE
Validate phone_number and recipient_identifier

### ISSUE THREAD

**Issue by k-macmillan** [department-of-veterans-affairs/va-enp-api#118]
## User Story - Business Need
Pydantic will allow us to validate phone number or recipient before getting into the handler. We should leverage that.

- [x] Sync with Kyle when ticket is picked up.
- [x] Ticket is understood, and QA has been contacted (if the ticket has a QA label).

### User Story(ies)

**As a** VA Notify dev
**I want** Pydantic to handle data entry validation
**So that** our code doesn't have to worry about it

### Additional Info and Resources

- Pydantic [custom validators](https://docs.pydantic.dev/latest/concepts/validators/)
- Phone number [validation](https://docs.pydantic.dev/latest/api/pydantic_extra_types_phone_numbers/)

## Acceptance Criteria

- [x] Only `phone_number` or `recipient_identifier` is used
- [x] Validation on `phone_number` ~(E.164 format)~ based on notification-api
- [x] Validation on `recipient_identifier` (matches expected formats)
- [x] This work is added to the sprint review slide deck (key win bullet point and demo slide)

## QA Considerations

Edge cases such as VOIP numbers.
- [x] Exercise the  POST `/v2/notifications/sms` constraints on the `phone_number` property
  - [x] must be a valid mobile number
 ~- [ ] must begin with +1~    - notification-api accepts numbers w/ +1 and so does ENP
- [x] Exercise the  POST `/v2/notifications/sms` constraints on the `recipient_identifier` property
  - [x] `id_type` should only be any of specified enum
  - [x] `id_value` should only be of the values accepted for each `id_type` [vaprofileid, edipi, birlsid, pid should be a string of numbers only, icn should be a string of numbers with the letter V at the -7 slice
 - [x] Exercise the  POST `legacy/v2/notifications/sms` constraints on the `phone_number` property
   - [x] must be a valid mobile number
  ~- [ ] must begin with +1~    - notification-api accepts numbers w/ +1 and so does ENP
- [x] Exercise the  POST `legacy/v2/notifications/sms` constraints on the `recipient_identifier` property
  - [x] `id_type` should only be any of specified enum
  - [x] `id_value` should only be of the values accepted for each `id_type` [vaprofileid, edipi, birlsid, pid should be a string of numbers only, icn should be a string of numbers with the letter V at the -7 slice

## Potential Dependencies

Dependent on #117

## Out of Scope
QA Automation to be handled in a separate ticket

## Dev Testing

Sample valid request in the log
![image](https://github.com/user-attachments/assets/3ba53185-4464-4bdd-bc9b-e9786164f8c5)

Sample invalid request in the log
![image](https://github.com/user-attachments/assets/5ec29a88-4dc7-41f4-91e9-b271dc3ac533)

Valid Recipient Identifier Formats

1. Valid BIRLSID
`{
  "template_id": "a71400e3-b2f8-4bd1-91c0-27f9ca7106a1",
  "sms_sender_id": "4f44ffc8-1ff8-4832-b1af-0b615691b6ea",
  "recipient_identifier": {
    "id_type": "BIRLSID",
    "id_value": "123456"
  }
}`

2. Valid EDIPI
`{
  "template_id": "a71400e3-b2f8-4bd1-91c0-27f9ca7106a1",
  "sms_sender_id": "4f44ffc8-1ff8-4832-b1af-0b615691b6ea",
  "recipient_identifier": {
    "id_type": "EDIPI",
    "id_value": "1235456"
  }
}`

3. Valid ICN
`{
  "template_id": "a71400e3-b2f8-4bd1-91c0-27f9ca7106a1",
  "sms_sender_id": "4f44ffc8-1ff8-4832-b1af-0b615691b6ea",
  "recipient_identifier": {
    "id_type": "ICN",
    "id_value": "1234567890V123456"
  }
}`

4. Valid PID
`{
  "template_id": "a71400e3-b2f8-4bd1-91c0-27f9ca7106a1",
  "sms_sender_id": "4f44ffc8-1ff8-4832-b1af-0b615691b6ea",
  "recipient_identifier": {
    "id_type": "PID",
    "id_value": "123456"
  }
}`

5. Valid VAPROFILEID
`{
  "template_id": "a71400e3-b2f8-4bd1-91c0-27f9ca7106a1",
  "sms_sender_id": "4f44ffc8-1ff8-4832-b1af-0b615691b6ea",
  "recipient_identifier": {
    "id_type": "VAPROFILEID",
    "id_value": "123456"
  }
}`

Valid Both recipient identifier and phone number provided

`{
  "template_id": "a71400e3-b2f8-4bd1-91c0-27f9ca7106a1",
  "sms_sender_id": "4f44ffc8-1ff8-4832-b1af-0b615691b6ea",
  "recipient_identifier": {
    "id_type": "BIRLSID",
    "id_value": "123456"
  },
  "phone_number": "+18005550101"
}`

Invalid Recipient Identifiers

1. Invalid id type
`{
  "template_id": "a71400e3-b2f8-4bd1-91c0-27f9ca7106a1",
  "sms_sender_id": "4f44ffc8-1ff8-4832-b1af-0b615691b6ea",
  "recipient_identifier": {
    "id_type": "INVALID",
    "id_value": "123456"
  }
}`

2. Invalid BIRLSID (non-numeric)
`{
  "template_id": "a71400e3-b2f8-4bd1-91c0-27f9ca7106a1",
  "sms_sender_id": "4f44ffc8-1ff8-4832-b1af-0b615691b6ea",
  "recipient_identifier": {
    "id_type": "BIRLSID",
    "id_value": "INVALID"
  }
}`

3. Invalid ICN (missing 'V')
`{
  "template_id": "a71400e3-b2f8-4bd1-91c0-27f9ca7106a1",
  "sms_sender_id": "4f44ffc8-1ff8-4832-b1af-0b615691b6ea",
  "recipient_identifier": {
    "id_type": "ICN",
    "id_value": "1234567890123456"
  }
}`

4. Invalid ICN (short prefix)
`{
  "template_id": "a71400e3-b2f8-4bd1-91c0-27f9ca7106a1",
  "sms_sender_id": "4f44ffc8-1ff8-4832-b1af-0b615691b6ea",
  "recipient_identifier": {
    "id_type": "ICN",
    "id_value": "123456789V123456"
  }
}`

5. Invalid ICN (short suffix)
`{
  "template_id": "a71400e3-b2f8-4bd1-91c0-27f9ca7106a1",
  "sms_sender_id": "4f44ffc8-1ff8-4832-b1af-0b615691b6ea",
  "recipient_identifier": {
    "id_type": "ICN",
    "id_value": "1234567890V12345"
  }
}`

Missing either recipient identifier or phone number
`{
  "template_id": "a71400e3-b2f8-4bd1-91c0-27f9ca7106a1",
  "sms_sender_id": "4f44ffc8-1ff8-4832-b1af-0b615691b6ea"
}`

**Comment by cris-oddball** [comment]
### QA Convo

- recipient_identifier includes VAPROFILEID, EDIPI, PID, ICN, BIRLSID
- must have one or the other, can have  both
- phone_numbers can be international
- recipient_identifier first looks up in MPI to get the VAPROFILEID; if we have the VAPROFILE then we call vaprofile for the contact info
- my outstanding question is how do we exercise these routes? Chris and Cris to jump on a call to validate these together

**Comment by ChrisJohnson-CDJ** [comment]
Working through existing notification-api validation tests and error messages. Building unit tests first in ENP (TDD) since ticket validation will rely on them.

**Comment by ChrisJohnson-CDJ** [comment]
Talked with Cris and Kyle to discuss how notification-api is currently accepting mixed format phone number strings and reformatting to E.164 instead of validating the input as strict E.164. The AC of the ticket has been updated to reflect that ENP needs to follow the same patterns for validating phone numbers.

```
('phone_number', 'description'),
    [
      ('+12025550123', 'E.164 format'),
      ('+1 202 555 0123', 'US country code with spaces'),
      ('+1(202)5550123', 'US country code with parenthesis'),
      ('+1-202-555-0123', 'US country code with hyphens'),
      ('+1 (202) 555-0123', 'US country code with mixed format'),
      ('2025550123', 'Local 10-digit'),
      ('202 555 0123', 'Local 10-digit with spaces'),
      ('202-555-0123', 'Local 10-digit with hyphens'),
      ('(202)5550123', 'Local 10-digit with parenthesis'),
      ('(202) 555-0123', 'Local 10-digit with mixed format'),
      ('12025550123', 'North American dialing'),
      ('1 202 555 0123', 'North American dialing with spaces'),
      ('1-202-555-0123', 'North American dialing with hyphens'),
      ('1(202)555-0123', 'North American dialing with parenthesis'),
      ('1 (202) 555-0123', 'North American dialing with mixed format'),
    ],
```

**Comment by ChrisJohnson-CDJ** [comment]
Created bug ticket to address notification-api phone number validation issue with vanity numbers (1-800-DIG-SAFE)

https://github.com/department-of-veterans-affairs/notification-utils/issues/208

**Comment by ChrisJohnson-CDJ** [comment]
While refactoring the existing validation code for use in ENP, I also discovered the the check for an invalid country prefix will never fail since phonenumbers.PhoneNumberMatcher would not return the number as valid in the first place.

**Comment by ChrisJohnson-CDJ** [comment]
After talking with Cris about testing, she pointed me to a possible alternate validation strategy to look into.

ValidatedPhoneNumber = Annotated[PhoneNumber, PhoneNumberValidator(default_region='US', number_format='E164')]

It passes all my notification-api test cases, both local and international (Philippines)

**Comment by ChrisJohnson-CDJ** [comment]
After discussing the pros/cons regarding rejecting vanity (non-numeric) phone numbers, the consensus from Kyle and Ksenia is to reject them. This does however require an additional validation check to be added onto the stock phone_numbers PhoneNumberValidator.

**Comment by cris-oddball** [comment]
PR approved and merged, cannot deploy until future work is finished.

Will being testing this Tuesday, 25 FEB 2025.

**Comment by cris-oddball** [comment]
### QA TESTING IN PROGRESS

Using the checklist in the description to exercise happy path and negative tests in swagger and checking logs. Posting anything to the notification-api db is out of scope for this ticket.

**Comment by cris-oddball** [comment]
The following tickets need to be created as a result of this testing:

- Story that error messages need to match the same format as notification-api
- Bug that only the upper case enumerated values are accepted for recipient_identifier

Also noting that QA was not previously able to validate the rest of the POST sms payloads from #117 and that testing is still outstanding.

**Comment by cris-oddball** [comment]
Closing ticket as Done
